src/cluster.c

```diff
@@ -713,7 +713,7 @@ void clusterAcceptHandler(aeEventLoop *el, int fd, void *privdata, int mask) {
 
     /* If the server is starting up, don't accept cluster connections:
      * UPDATE messages may interact with the database content. */
-    if (server.masterhost == NULL && server.loading) return;
+    if (server.masterhost == NULL && (server.loading || server.async_loading)) return;
```

Comment:
Thanks for the review and feedback @oranagra 
Sounds good with applying the behavior only when replica was previously in sync with a master. Could the check rely on master_link_status == up?
And what about the cases where we're running standalone and replica started from an unavailable master but managed to load it's own RDB from disk (so it's a master temporarily), then master comes up shortly and starts full sync?

Reply:
@eduardobr i gave it a lot of thought, here's what i think.

Here's a list of scenarios:
1. a replica gets disconnected a gets a full-sync when it reconnects
2. a replica was connected to one master and got redirected to a new master (of the same replication chain) and got a full sync for some reason.
3. redis was restarted from rdb file (after a graceful shutdown), and has replicaof config, attempted a partial sync with the replid it got from the rdb file, but gets a full sync.
4. the replica was a master till now and got some data from that era, then gets a REPLICAOF command and does a full-sync
5. the replica started empty and is attempting it's it's first sync ever.
6. the replica was replicating from a certain master, and is now full-syncing from another master that's completely unrelated to the previous one (i.e. imagine a sharded db with two shards, and two replicas, m1, m2, r1, r2, and now r2 that has the data of m2 is made a replica of r1, so it gonna get the data from m1. we don't want it to serve the data it has from the past).

i think the first 3 scenarios want so allow clients to read the data, but on the last 3 we don't!
i think that the rule here is that if the current dataset in the db represent a valid point in time of of the master're were full-syncing from, then we wanna allow reads, but if it doesn't represent a certain point in time of that master (possibly represents data of another master, or replication chain), then we don't want to serve reads.

I think the only way to tell that is if in `slaveTryPartialResynchronization` we get a full-sync with the same replid that we asked for (different offset since the offset we asked for doesn't exist in the backlog anymore).

the disadvantage is that there's one case that it'll miss: if the master at the head of the replication chain changed, and we switched the replid, then we reconnect and instead of getting successful partial-sync, we happen to get a full one. in this case we have no way to tell that in fact the old replid is actually an valid point in time of the new one.

this explanation is a bit long and messy, i hope i managed to convey the idea.

Reply:
Thinking about the case that is hard to deal with:

> "if the master at the head of the replication chain changed, and we switched the replid, then we reconnect and instead of getting successful partial-sync, we happen to get a full one."

I assume this is case 6.
When could that happen (that a replica will follow a master of another shard)? Is that only if some REPLICAOF command was raised?

Could we track if any REPLICAOF was executed since last successful replication? Then if there was any, we simply don't serve reads during replication. Sounds like a simple bit value that we set 1 when REPLICAOF is executed and set to 0 when successful replication finishes.

Reply:
the case i was describing last, which is not part of 1..6 is this:

* imagine a replication chain A <- B <- C (i.e. A is the master at the head of the chain)
* now normally, if B is promoted to be the new master, and A becomes a replica of B or C, what will happen is that B generates a new replid, and disconnects C. C then re-connects and succeeds a partial sync and acquires the new replid
* but if instead C takes too long to reconnect and the PSYNC fails and results in a full-sync, then it'll see that the replid it got is different than the one it asked for, and in our case (if we choose to take my suggestion in this comment), will avoid doing a swapdb based replication (or at least avoid serving traffic during the loading).

Lets call this case 7. it is different than 6 since in 6 we were talking of two different masters, not holding different points in time of the same data.

I think this compromise is ok, and i don't think case 7 is very likely, so i still think we can continue with the suggestion i gave in my comment.

@soloestoy @madolson @guybe7 @yossigo feel free to suggest a better idea.
the context is, how to decide in which cases to use swapdb based diskless repl and still serve clients during loading, and in which cases to avoid that (read my suggestion [here](https://github.com/redis/redis/pull/9323#discussion_r685941509))

Reply:
Applying this logic, when you have a setup with cluster disabled containing 1 master and some replicas and the master restarts, then we still won't serve reads during full sync because master will come up with a new replication id, but if it's not cluster and we know that no REPLICAOF was issued, isn't always the case it's safe to continue serving reads?

Reply:
@eduardobr you're right, but that's actually another bug. when the master restarts the replica shouldn't do full-sync: #8015
I think that counting REPLICAOF commands, or keeping track of whether or not the REPLICAOF command was with the same ip+port pair or different ones, is the wrong thing to do.
also note that it would also mean that when you switch to replicate from one source replica to another (in a replica chain), this will fail (we'll think we got connected to a new master), whereas my replid design will succeed.
in fact we have that replid mechanism for exactly that reason (to know which master/timeline we're part of).

We can even solve the case that i said is problematic (case 7), if we improve the protocol. i.e. some future PSYNC can use multiple replid+offset pairs when asking for sync, or and the PSYNC reply can carry multiple pairs back, and even carry part of the backlog. but we should lave that out of the scope now, since it requires protocol changes.

Reply:
Thanks for clarifying @oranagra 

> also note that it would also mean that when you switch to replicate from one source replica to another (in a replica chain), this will fail (we'll think we got connected to a new master), whereas my replid design will succeed

That suggestion about checking for an issued REPLICAOF would be only when replid is different in the handshake. A fallback logic (and only for cluster disabled if that helps). But I see the point to make a proper fix in the protocol.
Also, the PR mentioned relies on master loaded from RDB (I guess masters using RDB as primary persistence are not common anymore because that has high data loss potential and AOF is just available to be used, or I’m wrong?)

In summary, I’ll be glad to move on, but the original issue I thought to be solving wouldn’t be solved, even though there are still other improvements.

Drifting a bit, the scenario I’m trying to cover is to make redis more reliable when used as a primary database and not only as cache. I’m using a standalone setup to achieve virtually no data loss (having fsync=always in master for example). But it’s impossible to restart master in this scenario without putting down all replicas afaik (unless there’s some manual job to disconnect them and reconnect one by one). But in Kubernetes everything could happen.

Reply:
@eduardobr you're referring to graceful restart, right? (Not a crash recovery) 
See the comment I posted in https://github.com/redis/redis/pull/8015#issuecomment-897876142 I think maybe that can solve both of your problems, and also get you faster startup time. 

I don't like the idea of matching REPLICAOF commands and looking at their args. 
I do like to solve this problem for graceful restarts, even when AOF is configured. 
P. S. There is some [discussion](https://github.com/redis/redis/discussions/9282) about annotating AOF, but I still think it's not much good for graceful restarts, only crash recovery.

Reply:
I had a lengthy discussion about this with @yossigo.
we think the current plan is good.
i.e.
* If there's a PSYNC, there's no problem.
* If there's a FULL sync with the same replid, and the user uses repl-diskless-load=swapdb, we'll keep serving reads.
* If there's both a FULL sync and a replid change (either because of replication topology change or a master restart), we won't serve reads during sync.
* however, since we're gonna implement #8015, we don't expect FULL syncs after a graceful master restart.

we can proceed to implement this mechanism then.

Reply:
@oranagra I believe the last commit settles this strategy then. Added an extra test as well.